compiler/rustc_middle/src/ty/consts.rs

```diff
@@ -294,14 +275,88 @@ impl<'tcx> Const<'tcx> {
         Self::from_bits(tcx, n as u128, ParamEnv::empty().and(tcx.types.usize))
     }
 
-    /// Attempts to convert to a `ValTree`
-    pub fn try_to_valtree(self) -> Option<ty::ValTree<'tcx>> {
+    /// Returns the evaluated constant
+    pub fn eval(
+        self,
+        tcx: TyCtxt<'tcx>,
+        param_env: ParamEnv<'tcx>,
+        span: Option<Span>,
+    ) -> Result<ValTree<'tcx>, ErrorHandled> {
+        assert!(!self.has_escaping_bound_vars(), "escaping vars in {self:?}");
         match self.kind() {
-            ty::ConstKind::Value(valtree) => Some(valtree),
-            _ => None,
+            ConstKind::Unevaluated(unevaluated) => {
+                // HACK(eddyb) this erases lifetimes even though `const_eval_resolve`
+                // also does later, but we want to do it before checking for
+                // inference variables.
+                // Note that we erase regions *before* calling `with_reveal_all_normalized`,
+                // so that we don't try to invoke this query with
+                // any region variables.
+
+                // HACK(eddyb) when the query key would contain inference variables,
+                // attempt using identity args and `ParamEnv` instead, that will succeed
+                // when the expression doesn't depend on any parameters.
+                // FIXME(eddyb, skinny121) pass `InferCtxt` into here when it's available, so that
+                // we can call `infcx.const_eval_resolve` which handles inference variables.
+                let param_env_and = if (param_env, unevaluated).has_non_region_infer() {
+                    tcx.param_env(unevaluated.def).and(ty::UnevaluatedConst {
+                        def: unevaluated.def,
+                        args: GenericArgs::identity_for_item(tcx, unevaluated.def),
+                    })
+                } else {
+                    tcx.erase_regions(param_env)
+                        .with_reveal_all_normalized(tcx)
+                        .and(tcx.erase_regions(unevaluated))
+                };
```

Comment:
This is some magic that only the Valtree path has, maybe that's where the perf diff comes from?